Source/WebCore/platform/graphics/texmap/TextureMapperLayer3DRenderingContext.cpp

```diff
@@ -29,118 +29,266 @@
+
+    // Perform a broad-phase sweep-and-prune across x axis to identify potential intersections.
+    // By determining which layers might intersect, we can limit BSP cutting planes to those areas only,
+    // preventing unnecessary splitting of layers that are spatially distant.
+    // This optimization matters because the TextureMapper uses stencil operations to render split layers,
+    // and on some hardware, stencil usage is slow. Reducing unnecessary splits helps improve performance.
+    auto potentialIntersections = sweepAndPruneXAxis(layers);
+
+    // Determine if any layer pairs intersect on the ZY plane. An intersection implies that the rendering
+    // order is either ambiguous or overlapping, necessitating the use of a BSP tree for proper ordering.
+    bool hasIntersections = false;
+    if (!potentialIntersections.isEmpty()) {
+        for (const auto& p : potentialIntersections) {
+            auto quadA = projectLayerToZYPlane(layers[p.first]);
+            auto quadB = projectLayerToZYPlane(layers[p.second]);
+            if (quadsIntersect(quadA, quadB)) {
+                hasIntersections = true;
+                break;
             }
         }
-    });
+    }
+
+    // If no intersections are detected, the BSP tree building process is skipped and a fast path is taken.
+    // Other scenarios are not currently considered for optimization.
+    if (!hasIntersections) {
+        // Sort back to front
+        std::sort(layers.begin(), layers.end(), [](const Layer& layerA, const Layer& layerB) {
+            auto layerAminZ = layerA.geometry.bounds().min().z();
+            auto layerBminZ = layerB.geometry.bounds().min().z();
+            return layerAminZ < layerBminZ;
+        });
 
-    unsigned clipBufferSize = clipVertices.size() * sizeof(float);
-    auto clipBuffer = textureMapper.acquireBufferFromPool(clipBufferSize, TextureMapperGPUBuffer::Type::Vertex);
-    clipBuffer->updateData(clipVertices.data(), 0, clipBufferSize);
-
-    // Paint
-    traverseTree(*root, [&clipVertices, &clipBuffer, &paintLayerFunction](TextureMapperLayerNode& node) {
-        for (auto& polygon : node.polygons) {
-            unsigned numberOfClipVertices = polygon.isSplitted ? polygon.geometry.numberOfVertices() : 0;
-
-            Vector<FloatPoint> points;
-            if (numberOfClipVertices > 0) {
-                points.reserveCapacity(numberOfClipVertices);
-                auto xy = clipVertices.subvector(polygon.clipVertexBufferOffset, numberOfClipVertices * 2);
-                for (size_t i = 0; i < xy.size(); i += 2)
-                    points.append(FloatPoint(xy.at(i), xy.at(i + 1)));
+        for (auto& layer : layers)
+            paintLayerFunction(layer.textureMapperLayer, { });
+    } else {
+        Deque<Layer> layerDeque;
+        for (auto& layer : layers)
+            layerDeque.append(WTFMove(layer));
+        layers.clear();
```

Comment:
Can we use a Deque for the layers and then we don't need to convert the Vector to Deque here?

Reply:
sweepAndPruneXAxis operates with indices and returns index pairs. With Vector it's fast to access index directly. However, code has already change quite a bit from original approach and sweepAndPruneXAxis can be probably implemented a bit differently so that Deque can be used for the layers

Reply:
Seem the first part of the algorithm cannot be done with Deque. SweepAndPruneXAxis needs sort layers and thus random access iterator. Deque doesn't provide such.